Mathlib/Data/Finsupp/MonomialOrder/DegLex.lean

```diff
@@ -0,0 +1,261 @@
+
+@[simp] lemma DegLex.forall_iff {p : DegLex α → Prop} : (∀ a, p a) ↔ ∀ a, p (toDegLex a) := Iff.rfl
+@[simp] lemma DegLex.exists_iff {p : DegLex α → Prop} : (∃ a, p a) ↔ ∃ a, p (toDegLex a) := Iff.rfl
+
+noncomputable instance [AddCommMonoid α] :
+    AddCommMonoid (DegLex α) := ofDegLex.addCommMonoid
+
+theorem toDegLex_add [AddCommMonoid α] (a b : α) :
+    toDegLex (a + b) = toDegLex a + toDegLex b := rfl
+
+theorem ofDegLex_add [AddCommMonoid α] (a b : DegLex α) :
+    ofDegLex (a + b) = ofDegLex a + ofDegLex b := rfl
+
+namespace Finsupp
+
+/-- `Finsupp.DegLex r s` is the homogeneous lexicographic order on `α →₀ M`,
+where `α` is ordered by `r` and `M` is ordered by `s`.
+The type synonym `DegLex (α →₀ M)` has an order given by `Finsupp.DegLex (· < ·) (· < ·)`. -/
+protected def DegLex (r : α → α → Prop) (s : ℕ → ℕ → Prop) :
+    (α →₀ ℕ) → (α →₀ ℕ) → Prop :=
+  (Prod.Lex s (Finsupp.Lex r s)) on (fun x ↦ (x.degree, x))
+
+theorem degLex_def {r : α → α → Prop} {s : ℕ → ℕ → Prop} {a b : α →₀ ℕ} :
+    Finsupp.DegLex r s a b ↔ Prod.Lex s (Finsupp.Lex r s) (a.degree, a) (b.degree, b) :=
+  Iff.rfl
+
+theorem DegLex.wellFounded
+    {r : α → α → Prop} [IsTrichotomous α r] (hr : WellFounded (Function.swap r))
+    {s : ℕ → ℕ → Prop} (hs : WellFounded s) (hs0 : ∀ ⦃n⦄, ¬ s n 0) :
+    WellFounded (Finsupp.DegLex r s) := by
+  have wft := WellFounded.prod_lex hs (Finsupp.Lex.wellFounded' hs0 hs hr)
+  rw [← Set.wellFoundedOn_univ] at wft
+  unfold Finsupp.DegLex
+  rw [← Set.wellFoundedOn_range]
+  exact Set.WellFoundedOn.mono wft (le_refl _) (fun _ _ ↦ trivial)
+
+instance [LT α] : LT (DegLex (α →₀ ℕ)) :=
+  ⟨fun f g ↦ Finsupp.DegLex (· < ·) (· < ·) (ofDegLex f) (ofDegLex g)⟩
+
+theorem DegLex.lt_def [LT α] {a b : DegLex (α →₀ ℕ)} :
+    a < b ↔ (toLex ((ofDegLex a).degree, toLex (ofDegLex a))) <
+        (toLex ((ofDegLex b).degree, toLex (ofDegLex b))) :=
+  Iff.rfl
+
+theorem DegLex.lt_iff [LT α] {a b : DegLex (α →₀ ℕ)} :
+    a < b ↔ (ofDegLex a).degree < (ofDegLex b).degree ∨
+    (((ofDegLex a).degree = (ofDegLex b).degree) ∧ toLex (ofDegLex a) < toLex (ofDegLex b)) := by
+  simp only [Finsupp.DegLex.lt_def, Prod.Lex.lt_iff]
+
+variable [LinearOrder α]
+
+instance DegLex.isStrictOrder : IsStrictOrder (DegLex (α →₀ ℕ)) (· < ·) :=
+  { irrefl := fun a ↦ by simp [DegLex.lt_def]
+    trans := by
+      intro a b c hab hbc
+      simp only [DegLex.lt_iff] at hab hbc ⊢
+      rcases hab with (hab | hab)
+      · rcases hbc with (hbc | hbc)
+        · left; exact lt_trans hab hbc
+        · left; exact lt_of_lt_of_eq hab hbc.1
+      · rcases hbc with (hbc | hbc)
+        · left; exact lt_of_eq_of_lt hab.1 hbc
+        · right; exact ⟨Eq.trans hab.1 hbc.1, lt_trans hab.2 hbc.2⟩ }
+
+/-- The partial order on `Finsupp`s obtained by the homogeneous lexicographic ordering.
+See `Finsupp.DegLex.linearOrder` for a proof that this partial order is in fact linear. -/
+instance DegLex.partialOrder : PartialOrder (DegLex (α →₀ ℕ)) :=
```

Comment:
Then, you may define `DegLex.linearOrder` directly, and remove `DegLex.partialOrder`.

Reply:
I had guessed that it could be interesting — for mathlib's eventual future applications — to define all intermediate structures that depend on less assumptions, but that's indeed irrelevant to the one I have in mind.

Reply:
In any case, I removed the definition of the partial order.